docs/source/extension/extension_dual_compatibility.rst

```diff
@@ -0,0 +1,107 @@
+features that both Lab and Notebook 7 have, you can simply install it and it
+will seamlessly work in both Lab and Notebook 7.
+
+An extension that adds a single self-contained icon widget to the top bar
+of the UI, for instance, doesn't need to do anything at all to be compatible
+with both Lab and Notebook 7 (both apps have a top area that can hold the
+widget, so it will be visible in both Lab and Notebook 7 upon install and
+after launch).
+
+Not all extensions will be able to use "Do Nothing" compatibility: The
+"main" area in Lab is one example of a feature/UI area that is not supported
+by Notebook 7 where "Do Nothing" compatibility won't work.
+
+Compatibility in Complex Extensions
+
+Complex extensions can be adapted to work in both apps. The approach and
+effort needed to achieve compatibility depends on the design of your extension.
+
+Some features only exist in Lab, or in Notebook 7, and in those cases you will
+have some extra work to do to ensure compatibility, or to disable features that
+are not available in one of the interfaces. On top of that, you can specialize
+or add behaviors that are specific only to Lab or only to Notebook 7.
+
+With that in mind, here are some general tips for designing compatible extensions:
+
+- Organize your code: Identify which features in your extension are Lab-only
+  and which are Notebook 7-only, and try to separate those from the most
+  important core widgets and classes (you should put this app-specific code
+  in a different place than your main extension classes...if you include
+  app-specific code inside your main classes and widgets, it will be much
+  harder to achieve compatibility and pull those pieces out later)
+- Consider making some features optional, and enable them only when the
+  app-specificfeatures they need are available (more on this "optional-extras"
+  approach later)
+- Note that adding specialized app-specific extra behaviors and features
+  (features that rely on capabilities that only exist in either Lab or
+  Notebook 7) may require more complex compatibility architectures
+
+Solutions
+
+The "Do Nothing" Strategy
+
+Extensions that only use features common to both Lab and Notebook 7 don't
+need any modifications to work in both apps. Users who install your extension
+will be able to seamlessly run it in either Lab or Notebook 7.
+
+Keep in mind that some features for Lab are not available in Notebook 7:
+
+- The "main" area in Lab does not exist in Notebook 7
+- Notebook 7 does not have a status bar
+
+See `this example <https://www.youtube.com/watch?v=mqotG1MkHa4>`_ of a
+compatible top-bar-icon-widget that works in both Lab and Notebook 7
+out of the box.
+
+Using Multiple Plugins to Enable App Specific Behaviors
+
+One way to selectively enable app-specific features is to export a list of
+plugins from your extension, then use the "requires" plugin property to request
+Lab-only or Notebook-7 only features (testing for ILabShell or INotebookShell
```

Comment:
It would be better to not mention `INotebookShell` to prevent extensions authors from relying too much on it.

Also `INotebookShell` does not provide much than just the `currentChanged` signal compared to a `JupyterFront.IShell`.

Reply:
Instead we can mention how widgets can be positioned in other areas than the `main` area in both JupyterLab and Notebook 7:

- https://jupyterlab.readthedocs.io/en/latest/user/interface_customization.html
- https://jupyter-notebook.readthedocs.io/en/latest/configuring/interface_customization.html

And expand the documentation there if needed.

Reply:
I left some notes about INotebookShell and tailoring frontend-specific behaviors above (need to give some warnings but I think still valid in summary).

Other areas I can definitely discuss.

Reply:
These should be resolved now.